Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,310 @@
+            if (doc.AbsoluteExpiration < now)
+            {
+                return null;
+            }
+            var slidingExpiration = doc.LastRefresh.GetValueOrDefault()
+                + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.GetValueOrDefault());
+            if (slidingExpiration < now)
+            {
+                return null;
+            }
+            return doc.Value;
+        }
+
+        void IDistributedCache.Refresh(string key)
+        {
+            try
+            {
+                Task.Run(() => _cacheEntries.Document(key)
+                    .UpdateAsync("LastRefresh", _clock.GetCurrentDateTimeUtc())).Wait();
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        async Task IDistributedCache.RefreshAsync(string key, CancellationToken token)
+        {
+            try
+            {
+                await _cacheEntries.Document(key).UpdateAsync(
+                    "LastRefresh", _clock.GetCurrentDateTimeUtc(), cancellationToken: token)
+                    .ConfigureAwait(false);
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        void IDistributedCache.Remove(string key) =>
+            Task.Run(() => _cacheEntries.Document(key).DeleteAsync()).Wait();
+
+        Task IDistributedCache.RemoveAsync(string key, CancellationToken token) =>
+            _cacheEntries.Document(key).DeleteAsync(cancellationToken: token);
+
+        /// <summary>
+        /// Creates a Firestore document that stores the cache value and its
+        /// expiration info, but does not store it in Firestore.
+        /// </summary>
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        private CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = _clock.GetCurrentDateTimeUtc(),
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds = options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = TimeSpan.FromDays(365000);  // 1000 years.

```

Comment:
You might want to make this an internal static readonly field, so you can use it in unit tests.
In fact, you could make the field the TotalSeconds value instead of a TimeSpan - but still use TimeSpan to derive it.

Reply:
Done.